modules/infinite-scroll/infinity.php

```diff
@@ -1679,6 +1692,182 @@ public function add_mejs_config( $scripts_data ) {
+
+			add_action( 'template_redirect', array( $this, 'amp_start_output_buffering' ), 0 );
+			add_action( 'shutdown', array( $this, 'amp_output_buffer' ), 1 );
+
+			if ( is_callable( "amp_{$template}_hooks" ) ) {
+				call_user_func( "amp_{$template}_hooks" );
+			}
+
+			// Warms up the amp next page markup.
+			// This should be done outside the output buffering callback started in the template_redirect.
+			$this->amp_get_footer_template();
+		}
+	}
+
+	/**
+	 * Start the AMP output buffering.
+	 *
+	 * @return void
+	 */
+	public function amp_start_output_buffering() {
+		ob_start( array( $this, 'amp_finish_output_buffering' ) );
+	}
+
+	/**
+	 * Flush the AMP output buffer.
+	 *
+	 * @return void
+	 */
+	public function amp_output_buffer() {
+		if ( ob_get_contents() ) {
+			ob_end_flush();
+		}
+	}
+
+	/**
+	 * Filter the AMP output buffer contents.
+	 *
+	 * @param string $buffer Contents of the output buffer.
+	 *
+	 * @return string|false
+	 */
+	public function amp_finish_output_buffering( $buffer ) {
+		// Hide WordPress admin bar on next page load.
+		$buffer = preg_replace(
+			'/id="wpadminbar"/',
+			'$0 next-page-hide',
+			$buffer
+		);
+
+		// Get the theme footers.
+		$footers = apply_filters( 'jetpack_amp_infinite_footers', array(), $buffer );
+
+		// Hook for themes to filter custom markup on next page load.
+		$buffer = apply_filters( 'jetpack_amp_infinite_output', $buffer );
+
+		// Add the amp next page markup.
+		$buffer = preg_replace(
+			'~</body>~',
+			$this->amp_get_footer_template( $footers ) . '$0',
+			$buffer
+		);
+
+		return $buffer;
+	}
+
+	/**
+	 * Get AMP next page markup with the custom footers.
+	 *
+	 * @param string[] $footers The theme footers.
+	 *
+	 * @return string
+	 */
+	protected function amp_get_footer_template( $footers = array() ) {
+		static $template = null;
+
+		if ( null === $template ) {
+			$template = $this->amp_footer_template();
+		}
+
+		if ( empty( $footers ) ) {
+			return $template;
+		}
+
+		return preg_replace(
+			'/%%footer%%/',
+			implode( '', $footers ),
+			$template
+		);
+	}
+
+	/**
+	 * AMP Next Page markup.
+	 *
+	 * @return string
+	 */
+	protected function amp_footer_template() {
+		ob_start();
+		?>
+<amp-next-page max-pages="<?php echo esc_attr( $this->amp_get_max_pages() ); ?>">
+	<script type="application/json">
+		[
+			<?php echo wp_json_encode( $this->amp_next_page() ); ?>
+		]
+	</script>
+	<div separator>
+		<?php echo wp_kses_post( apply_filters( 'jetpack_amp_infinite_separator', '' ) ); ?>
+	</div>
+	<div recommendation-box class="recommendation-box">
+		<template type="amp-mustache">
+			{{#pages}}
+			<?php echo wp_kses_post( apply_filters( 'jetpack_amp_infinite_older_posts', '' ) ); ?>
+			{{/pages}}
+		</template>
+	</div>
+	<div footer>
+		%%footer%%
+	</div>
+</amp-next-page>
+		<?php
+		return ob_get_clean();
+	}
+
+	/**
+	 * Get the AMP next page information.
+	 *
+	 * @return array
+	 */
+	protected static function amp_next_page() {
```

Comment:
Is there a reason this is static? It's being called on a object in line 1807

Reply:
@MattGeri 

That's a great catch.
For some reason, it made sense at some point. After the cleanup, I missed rethinking this one.